cs/src/core/protocols/CompactBinary.cs

```diff
@@ -148,6 +155,27 @@ public CompactBinaryWriter(O output, ushort version = 1)
         {
             this.output = output;
             this.version = version;
+            if (version == 2)
+            {
+                this.lengths = new List<uint>();
+                this.firstPassWriter = new CompactBinaryCounter(lengths);
+#if DEBUG
```

Comment:
I think that moving the lengthCheck code (other than the member variable) into conditional methods will result in cleaner code.

```
public CompactBinaryWriter(O output, ushort version = 1)
{
    this.output = output;
    this.version = version;
    if (version == 2)
    {
        this.lengths = new List<uint>();
        this.firstPassWriter = new CompactBinaryCounter(lengths);
    }
    else
    {
        this.lengths = null;
        this.firstPassWriter = null;
    }

    InitLengthCheck(version);
}

[Conditional(DEBUG)]
void InitLengthCheck(int version)
{
    if (version == 2)
    {
        this.lengthCheck = new Stack<long>();
    }
   ...
}
```
and similarly for `PushLengthCheck` and `PopLengthCheck`.